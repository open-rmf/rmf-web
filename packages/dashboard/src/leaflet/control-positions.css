/* FIXME: Check for changes when updating leaflet */
.leaflet-vertical-center {
  position: absolute;
  z-index: 1000;
  pointer-events: none;
	top: 50%;
	transform: translateY(-50%);
	-ms-transform: translateY(-50%);
	-webkit-transform: translateY(-50%);
	-o-transform: translateY(-50%);
	-moz-transform: translateY(-50%);
}
.leaflet-horizontal-center {
	left: 50%;
	transform: translateX(-50%);
	-ms-transform: translateX(-50%);
	-webkit-transform: translateX(-50%);
	-o-transform: translateX(-50%);
	-moz-transform: translateX(-50%);
}
.leaflet-vertical-center .leaflet-horizontal-center {
	transform: translate(-50%,-50%) !important;
	-ms-transform: translate(-50%,-50%) !important;
	-webkit-transform: translate(-50%,-50%) !important;
	-o-transform: translate(-50%,-50%) !important;
	-moz-transform: translate(-50%,-50%) !important;
}
.leaflet-horizontal-center .leaflet-control {
  float: none;
  display: inline-block;
}