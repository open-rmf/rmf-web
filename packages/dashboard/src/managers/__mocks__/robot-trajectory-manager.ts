import Debug from 'debug';
import {
  RobotTrajectoryManager,
  TimeRequest,
  TimeResponse,
  TrajectoryRequest,
  TrajectoryResponse,
} from '../robot-trajectory-manager';
import trajectories from './trajectories.json';

const debug = Debug('FakeTrajectoryManager');

export default class FakeTrajectoryManager implements RobotTrajectoryManager {
  async latestTrajectory(request: TrajectoryRequest): Promise<TrajectoryResponse> {
    debug('sending trajectory');
    if (request.param.map_name === 'L1') {
      const traj = trajectories[this.currentTraj++];
      this.currentTraj %= trajectories.length;
      // "deep clone" object
      return JSON.parse(JSON.stringify(traj)) as any;
    }

    return {
      response: 'trajectory',
      values: [],
      conflicts: [],
    };
  }

  serverTime(request: TimeRequest): Promise<TimeResponse> {
    throw new Error('Method not implemented.');
  }

  private currentTraj = 0;
}
