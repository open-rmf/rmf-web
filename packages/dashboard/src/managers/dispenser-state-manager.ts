import * as RomiCore from '@osrf/romi-js-core-interfaces';
import EventEmitter from 'eventemitter3';

type Events = {
  updated: [];
};

export default class DispenserStateManager extends EventEmitter<Events> {
  dispenserStates: Record<string, RomiCore.DispenserState> = {};

  startSubscription(transport: RomiCore.Transport) {
    this._subscriptions.push(
      transport.subscribe(RomiCore.dispenserStates, (dispenserState) => {
        this.dispenserStates[dispenserState.guid] = dispenserState;
        this.emit('updated');
      }),
    );
  }
  stopAllSubscriptions(): void {
    this._subscriptions.forEach((sub) => sub.unsubscribe());
  }

  private _subscriptions: RomiCore.Subscription[] = [];
}
