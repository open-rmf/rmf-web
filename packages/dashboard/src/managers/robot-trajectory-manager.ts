import { Knot } from 'react-components';
import { Authenticator } from 'rmf-auth';
import TrajectorySocketManager from './trajectory-socket-manager';

// RawVelocity received from server is in this format (x, y, theta)
export type RawVelocity = [number, number, number];

// RawPose2D received from server is in this format (x, y, theta)
export type RawPose2D = [number, number, number];

export interface RawKnot {
  t: number; // milliseconds
  v: RawVelocity;
  x: RawPose2D;
}

export interface TrajectoryRequest {
  request: 'trajectory';
  param: {
    map_name: string;
    duration: number;
    trim: boolean;
  };
  token?: string;
}

export interface TimeRequest {
  request: 'time';
  param: {};
  token?: string;
}

export interface TimeResponse {
  response: 'time';
  values: [number];
}

export interface Trajectory {
  id: number;
  shape: string;
  dimensions: number;
  segments: RawKnot[];
  robot_name: string;
  fleet_name: string;
  map_name: string;
}

export interface TrajectoryResponse {
  response: 'trajectory';
  values: Trajectory[];
  conflicts: Conflict[];
  error?: string;
}

export type Conflict = number[];

export interface RobotTrajectoryManager {
  serverTime(request: TimeRequest): Promise<TimeResponse>;
  latestTrajectory(request: TrajectoryRequest): Promise<TrajectoryResponse>;
}

interface Request {
  request: string;
  param: unknown;
}

interface Response {
  response: string;
  values: unknown;
  error?: string;
}

export class DefaultTrajectoryManager extends TrajectorySocketManager {
  constructor(ws: WebSocket, authenticator?: Authenticator) {
    super();
    if (ws) this._webSocket = ws;
    if (authenticator) this._authenticator = authenticator;
  }

  async latestTrajectory(request: TrajectoryRequest): Promise<TrajectoryResponse> {
    await this._authenticator?.refreshToken();
    const event = await this.send(JSON.stringify(request), this._webSocket);
    const resp = JSON.parse(event.data);
    const validResp = this._checkResponse(request, resp);
    if (resp.values === null || !validResp) {
      resp.values = [];
      resp.conflicts = [];
    }
    return resp as TrajectoryResponse;
  }

  async serverTime(request: TimeRequest): Promise<TimeResponse> {
    const event = await this.send(JSON.stringify(request), this._webSocket);
    const resp = JSON.parse(event.data);
    this._checkResponse(request, resp);
    return resp as TimeResponse;
  }

  static getRobotNameFromPathId(
    pathId: number,
    trajectories: readonly Trajectory[],
  ): string | undefined {
    const traj = trajectories.find((trajectory) => trajectory.id === pathId);
    return traj?.robot_name;
  }

  private _checkResponse(request: Request, resp: Response): boolean {
    if (resp.error) {
      if (resp.error === 'token expired') this._authenticator?.logout();
      else {
        throw new Error(resp.error);
      }
    } else if (request.request !== resp.response) {
      console.warn(
        `received response for wrong request. Request: ${request.request} Response: ${resp.response}`,
      );
      return false;
    }
    return true;
  }

  private _webSocket: WebSocket | undefined;
  private _authenticator?: Authenticator;
}

export function rawKnotsToKnots(rawKnots: RawKnot[]): Knot[] {
  const knots: Knot[] = [];

  for (const rawKnot of rawKnots) {
    const [poseX, poseY, poseTheta] = rawKnot.x;
    const [velocityX, velocityY, velocityTheta] = rawKnot.v;
    knots.push({
      pose: {
        x: poseX,
        y: poseY,
        theta: poseTheta,
      },
      velocity: {
        x: velocityX,
        y: velocityY,
        theta: velocityTheta,
      },
      time: rawKnot.t,
    });
  }

  return knots;
}
