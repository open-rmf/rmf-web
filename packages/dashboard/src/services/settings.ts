export type ThemeMode = 'default' | 'dark';

export interface Settings {
  themeMode: ThemeMode;
  microAppSettings: { [k: string]: unknown };
}

export function saveSettings(settings: Settings): void {
  localStorage.setItem('settings', JSON.stringify(settings));
}

export function loadSettings(): Settings {
  const settingsStr = localStorage.getItem('settings');
  if (!settingsStr) {
    return defaultSettings;
  }
  return JSON.parse(settingsStr);
}

const defaultSettings: Settings = {
  themeMode: 'default',
  microAppSettings: {},
};
