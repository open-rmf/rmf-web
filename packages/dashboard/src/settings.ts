export enum TrajectoryAnimation {
  None,
  Follow,
}

export enum ThemeMode {
  Light,
  Dark,
}

export enum UseTheme {
  True,
  False,
}

export interface Settings {
  trajectoryAnimation: TrajectoryAnimation;
  themeMode: ThemeMode;
  useTheme: UseTheme;
}

export function saveSettings(settings: Settings): void {
  localStorage.setItem('settings', JSON.stringify(settings));
}

export function loadSettings(): Settings {
  const settingsStr = localStorage.getItem('settings');
  if (!settingsStr) {
    return defaultSettings();
  }
  return JSON.parse(settingsStr);
}

export function defaultSettings(): Settings {
  return {
    trajectoryAnimation: TrajectoryAnimation.Follow,
    themeMode: ThemeMode.Light,
    useTheme: UseTheme.False,
  };
}
