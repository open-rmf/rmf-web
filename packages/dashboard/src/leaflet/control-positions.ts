import * as L from 'leaflet';

import './control-positions.css';

export type ControlPosition = (
  L.ControlPosition | 'topcenter' | 'bottomcenter' | 'leftcenter' | 'rightcenter' | 'center'
);

export function extendControlPositions() {
  // Implement feature from this PR: https://github.com/Leaflet/Leaflet/pull/5554
  const origInitControlPos = (L.Map as any).prototype._initControlPos;
  L.Map.include({
    _initControlPos() {
      origInitControlPos.call(this);
      this._controlCorners['topcenter'] = L.DomUtil.create(
        'div',
        'leaflet-top leaflet-horizontal-center',
        this._controlContainer,
      );
      this._controlCorners['bottomcenter'] = L.DomUtil.create(
        'div',
        'leaflet-bottom leaflet-horizontal-center',
        this._controlContainer,
      );
      this._controlCorners['leftcenter'] = L.DomUtil.create(
        'div',
        'leaflet-left leaflet-vertical-center',
        this._controlContainer,
      );
      this._controlCorners['rightcenter'] = L.DomUtil.create(
        'div',
        'leaflet-right leaflet-vertical-center',
        this._controlContainer,
      );
      this._controlCorners['center'] = L.DomUtil.create(
        'div',
        'leaflet-horizontal-center leaflet-vertical-center',
        this._controlContainer,
      );
    },
  });
}