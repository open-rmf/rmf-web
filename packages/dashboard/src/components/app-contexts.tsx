import React from 'react';
import ResourceManager from '../managers/resource-manager';
import { defaultSettings, Settings } from '../settings';

/* Declares the ResourcesContext which contains the resources used on the app*/
export const ResourcesContext = React.createContext<ResourceManager | undefined>(undefined);

export const SettingsContext = React.createContext(defaultSettings());

export interface AppController {
  setSettings: React.Dispatch<React.SetStateAction<Settings>>;
  saveSettings(settings: Settings): void;
  showSettings(show: boolean): void;
  toggleSettings(): void;
  showHelp(show: boolean): void;
  toggleHelp(): void;
  showHotkeysDialog(show: boolean): void;
  toggleHotkeysDialog(): void;
  /**
   * FIXME: Move this to settings, this sets if we should show the tooltip icon, and not trigger
   * a tooltip to show.
   */
  showTooltips(show: boolean): void;
  /**
   * FIXME: Move this to settings, this sets if we should show the tooltip icon, and not trigger
   * a tooltip to show.
   */
  toggleTooltips(): void;
}

export interface Tooltips {
  showTooltips: boolean;
}

export const TooltipsContext = React.createContext<Tooltips>({
  showTooltips: true,
});

export const AppControllerContext = React.createContext<AppController>({
  setSettings: () => {},
  saveSettings: () => {},
  showSettings: () => {},
  toggleSettings: () => {},
  showHelp: () => {},
  toggleHelp: () => {},
  showHotkeysDialog: () => {},
  toggleHotkeysDialog: () => {},
  showTooltips: () => {},
  toggleTooltips: () => {},
});

export interface AppContent {
  tabNames: string[];
}
export const AppContentContext = React.createContext<AppContent>({ tabNames: ['Building'] });
