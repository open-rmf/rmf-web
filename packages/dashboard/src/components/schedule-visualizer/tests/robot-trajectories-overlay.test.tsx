import { createMount } from '@material-ui/core/test-utils';
import * as RomiCore from '@osrf/romi-js-core-interfaces';
import { ReactWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as L from 'leaflet';
import React from 'react';
import { robotHash } from 'react-components';
import { act } from 'react-dom/test-utils';
import { Map as LMap } from 'react-leaflet';
import { Conflict, Trajectory } from '../../../managers/robot-trajectory-manager';
import FakeTrajectoryManager from '../../../managers/__mocks__/robot-trajectory-manager';
import { defaultSettings, TrajectoryAnimation } from '../../../settings';
import { SettingsContext } from '../../app-contexts';
import RobotTrajectoriesOverlay from '../robot-trajectories-overlay';

const mount = createMount();
const mapBound = new L.LatLngBounds([0, 25.794363144785166], [-17.53525484725833, 0]);

const robots: Record<string, RomiCore.RobotState> = {
  [robotHash('tinyRobot1', 'tinyRobot')]: {
    name: 'tinyRobot1',
    battery_percent: 100,
    location: { level_name: 'L1', x: 0, y: 0, yaw: 0, t: { sec: 0, nanosec: 0 } },
    mode: { mode: RomiCore.RobotMode.MODE_IDLE },
    model: 'tinyRobot',
    path: [],
    task_id: '',
  },
  [robotHash('tinyRobot2', 'tinyRobot')]: {
    name: 'tinyRobot2',
    battery_percent: 100,
    location: { level_name: 'L1', x: 0, y: 0, yaw: 0, t: { sec: 0, nanosec: 0 } },
    mode: { mode: RomiCore.RobotMode.MODE_IDLE },
    model: 'tinyRobot',
    path: [],
    task_id: '',
  },
};

const createWrapper = async (
  bounds: L.LatLngBoundsExpression,
  conflicts: Conflict[],
  robots: Record<string, RomiCore.RobotState>,
  trajs: Trajectory[],
) => {
  const settings = defaultSettings();
  // animations use apis not supported in jsdom
  settings.trajectoryAnimation = TrajectoryAnimation.None;
  let wrapper: ReactWrapper;
  await act(
    async () =>
      (wrapper = mount(
        <LMap>
          <SettingsContext.Provider value={settings}>
            <RobotTrajectoriesOverlay
              bounds={bounds}
              robots={robots}
              trajectories={trajs}
              conflicts={conflicts}
            />
          </SettingsContext.Provider>
        </LMap>,
      )),
  );
  return wrapper!;
};

describe('RobotTrajectoriesOverlay', () => {
  let trajMgr;
  let trajectoryData;
  let trajectoryValue: Trajectory;
  let trajectoryConflict: Conflict[];

  beforeEach(async () => {
    trajMgr = new FakeTrajectoryManager();
    trajectoryData = await trajMgr.latestTrajectory({
      request: 'trajectory',
      param: {
        map_name: 'L1',
        duration: 1000,
        trim: true,
      },
    });

    trajectoryValue = trajectoryData.values[0];
    trajectoryConflict = trajectoryData.conflicts;
  });

  it('renders without crashing', async () => {
    const wrapper = await createWrapper(mapBound, trajectoryConflict, robots, [trajectoryValue]);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
