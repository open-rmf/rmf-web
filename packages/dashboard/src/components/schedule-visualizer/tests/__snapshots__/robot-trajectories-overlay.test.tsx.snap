// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RobotTrajectoriesOverlay renders without crashing 1`] = `
<Map>
  <div />
</Map>
`;
